parser/src/input/mod.rs

```diff
@@ -0,0 +1,107 @@
+mod encoding;
+
+use crate::config::EncodingRef;
+use bytes::{Buf, BufMut, Bytes, BytesMut};
+use std::fs::File;
+use std::io::{self, Read, Seek};
+
+pub use self::encoding::{detect_encoding, TranscodingReader};
+
+/// Type of content input provided to parsers.
+pub enum Input {
+    File(File),
+    Stream(Box<dyn io::Read>),
+    //BufferedStream(Box<dyn io::BufRead>),
+}
+
+impl Input {
+    pub fn into_buffered_stream(self, buffer_size: usize) -> Box<dyn io::BufRead> {
+        match self {
+            Input::File(f) => Box::new(io::BufReader::with_capacity(buffer_size, f)),
+            //Input::BufferedStream(bs) => bs,
+            Input::Stream(s) => Box::new(io::BufReader::with_capacity(buffer_size, s)),
+        }
+    }
+
+    pub fn into_stream(self) -> Box<dyn io::Read> {
+        match self {
+            Input::File(f) => Box::new(f),
+            //Input::BufferedStream(bs) => bs as Box<dyn io::Read>,
+            Input::Stream(s) => s,
+        }
+    }
+
+    /// Reads at most `max_bytes` into memory, and returns an `Input` that is still positioned at
+    /// the beginning of the stream. The returned `Input` can be read as normal without missing any
+    /// bytes. The returned `Bytes` will contain a duplicate of the first `max_bytes` (at most)
+    /// from `self`.
+    pub fn peek_at(self, max_bytes: usize) -> io::Result<(Bytes, Input)> {
```

Comment:
nit: `peek_at` naming is confusing to me (what's the "at"? why not just `peek` ?)